### ISSUE TITLE
Missing binaries error in production

### ISSUE THREAD

**Issue by creativetags** [toy/image_optim_pack#12]
I've asked this [on StackOverflow](https://stackoverflow.com/questions/45611537/missing-binaries-error-with-image-optim-pack-in-production) but maybe it's more generic than just my own issue?

I'm trying to deploy a Rails app with paperclip-optimizer and it's giving me an error about missing binaries. I have bundled image_optim_pack to provide the binaries but still get this:

    ImageOptim::BinResolver::Error occurred
    pngcrush worker: `pngcrush` not found; please provide proper binary or di=
    sable this worker (--no-pngcrush argument or `:pngcrush =3D> false` throu=
    gh options)
    advpng worker: `advpng` not found; please provide proper binary or disabl=
    e this worker (--no-advpng argument or `:advpng =3D> false` through optio=
    ns)
    optipng worker: `optipng` not found; please provide proper binary or disa=
    ble this worker (--no-optipng argument or `:optipng =3D> false` through o=
    ptions)
    pngquant worker: `pngquant` not found; please provide proper binary or di=
    sable this worker (--no-pngquant argument or `:pngquant =3D> false` throu=
    gh options)
    jhead worker: `jhead` not found, `jpegtran` not found; please provide pro=
    per binary or disable this worker (--no-jhead argument or `:jhead =3D> fa=
    lse` through options)
    jpegoptim worker: `jpegoptim` not found; please provide proper binary or =
    disable this worker (--no-jpegoptim argument or `:jpegoptim =3D> false` t=
    hrough options)
    jpegtran worker: `jpegtran` not found; please provide proper binary or di=
    sable this worker (--no-jpegtran argument or `:jpegtran =3D> false` throu=
    gh options)
    gifsicle worker: `gifsicle` not found; please provide proper binary or di=
    sable this worker (--no-gifsicle argument or `:gifsicle =3D> false` throu=
    gh options)

- Server Ubuntu 14.04.5 LTS
- Rails 5.1.4
- Ruby 2.4.1
- nginx 1.9.10
- passenger 5.0.24
- image_optim (0.25.0)
- image_optim_pack (0.5.0.20170831 x86_64-linux)

**Comment by toy** [comment]
Please run `bundle exec image_optim --info` and see if image_optim_pack succeeded and if bins except `pngout` and `svgo` are resolved to those provided by image_optim_pack

**Comment by creativetags** [comment]
It gave me this:

```
image_optim v0.25.0
config:
  verbose: true
nice: 10
threads: 2
pack: true
skip_missing_workers: true
allow_lossy: false
cache_dir: 
cache_worker_digests: false
image_optim_pack: advpng, gifsicle, jhead, jpeg-recompress, jpegoptim, jpegtran, optipng, pngcrush from /home/ctgs/shared/bundle/ruby/2.4.0/gems/image_optim_pack-0.5.0.20170831-x86_64-linux/vendor/linux-x86_64 failed
No pack for this OS and/or ARCH, check verbose output
pngcrush worker: `pngcrush` not found; please provide proper binary or disable this worker (--no-pngcrush argument or `:pngcrush => false` through options)
pngout worker: `pngout` not found; please provide proper binary or disable this worker (--no-pngout argument or `:pngout => false` through options)
advpng worker: `advpng` not found; please provide proper binary or disable this worker (--no-advpng argument or `:advpng => false` through options)
optipng worker: `optipng` not found; please provide proper binary or disable this worker (--no-optipng argument or `:optipng => false` through options)
pngquant worker: `pngquant` not found; please provide proper binary or disable this worker (--no-pngquant argument or `:pngquant => false` through options)
jhead worker: `jhead` not found, `jpegtran` not found; please provide proper binary or disable this worker (--no-jhead argument or `:jhead => false` through options)
jpegoptim worker: `jpegoptim` not found; please provide proper binary or disable this worker (--no-jpegoptim argument or `:jpegoptim => false` through options)
jpegtran worker: `jpegtran` not found; please provide proper binary or disable this worker (--no-jpegtran argument or `:jpegtran => false` through options)
gifsicle worker: `gifsicle` not found; please provide proper binary or disable this worker (--no-gifsicle argument or `:gifsicle => false` through options)
svgo worker: `svgo` not found; please provide proper binary or disable this worker (--no-svgo argument or `:svgo => false` through options)
Workers by format:
```

**Comment by toy** [comment]
A bit weird that `image_optim_pack` warning doesn't mention `pngquant`.
Can you please check if the binaries in that vendor subfolder are executable?
I have a feeling that something may be wrong with standard libraries, can you also run ldd on those binaries?

**Comment by creativetags** [comment]
Here's the directory listing:
```
total 2.3M
drwxrwxr-x 2 ctgs ctgs 4.0K Sep 11 11:28 ./
drwxrwxr-x 3 ctgs ctgs 4.0K Sep 11 11:28 ../
-rwxr-xr-x 1 ctgs ctgs 344K Sep 11 11:28 advpng*
-rwxr-xr-x 1 ctgs ctgs 224K Sep 11 11:28 gifsicle*
-rwxr-xr-x 1 ctgs ctgs  77K Sep 11 11:28 jhead*
-rwxr-xr-x 1 ctgs ctgs  39K Sep 11 11:28 jpegoptim*
-rwxr-xr-x 1 ctgs ctgs 535K Sep 11 11:28 jpeg-recompress*
-rwxr-xr-x 1 ctgs ctgs  47K Sep 11 11:28 jpegtran*
-rwxr-xr-x 1 ctgs ctgs 300K Sep 11 11:28 libjpeg.so*
-rwxr-xr-x 1 ctgs ctgs 249K Sep 11 11:28 libpng.so*
-rwxr-xr-x 1 ctgs ctgs 112K Sep 11 11:28 libz.so*
-rwxr-xr-x 1 ctgs ctgs 101K Sep 11 11:28 optipng*
-rwxr-xr-x 1 ctgs ctgs 115K Sep 11 11:28 pngcrush*
-rwxr-xr-x 1 ctgs ctgs  84K Sep 11 11:28 pngquant*
```
ldd on any of those returns `not a dynamic executable`

**Comment by toy** [comment]
If I understand correctly it is either caused by 64 bit binaries on 32 bit system or shared libraries problem. What does `uname -i` output for you?

**Comment by toy** [comment]
Can you then please try to check what output will you get when running one of not working binaries and what for pngquant?

**Comment by creativetags** [comment]
When I try to run one of those I just get
`./jhead: No such file or directory`
pngquant gives the same

**Comment by toy** [comment]
Are you sure that you are running it from gem vendor/linux* directory? Can you also try using full path?

**Comment by creativetags** [comment]
Yeah, it's weird. I did the same using the full path. I think there might be something going on with the 64 vs 32 bit thing. I might have a 64-bit kernel and 32-bit environment? If I run `file /usr/bin/file` I get `/usr/bin/file: ELF 32-bit LSB  executable, Intel 80386, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.24, BuildID[sha1]=0428d646fd1bffff284d958f48e0d76a8f90f3e1, stripped`

**Comment by toy** [comment]
Can you try to use either linux 32 bit specific gem or the one without platform?

**Comment by creativetags** [comment]
Sure, where do I find those? I can't see them in your repos

**Comment by toy** [comment]
Every version of gem has variants with different platform. It seems adding platform option in gemfile works: `gem 'image_optim_pack', platform: 'ruby'`. Or you can remove other versions of gem and manually install platform specific one: `gem install image_optim_pack --platform=i686-linux` and bundle should pick that one.
Also it is interesting what does `gem environment` output in section "RUBYGEMS PLATFORMS" on your server and what does `bundle platform` output (the values are probably related)? It seems that your issue arrises from the fact that rubygems expect your system to be 64 bit, but it is not.

**Comment by creativetags** [comment]
Thanks for your help. I haven't managed to get it to work and am probably going to give up as it's taking too much time. I'm contemplating reinstalling Ubuntu to get everything 64-bit for simplicity in the future.

gem environment
RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-linux

bundle platform
Your platform is: x86_64-linux

**Comment by toy** [comment]
When creating platform specific gems I knew that there may be some problems, as too my knowledge bundler doesn't give any way of selecting which platform specific version to use, but I wanted to considerably reduce the gem size.
Normally if you have some gem already installed (and gem command allows you to specify platform) and bundler is allowed to use system gems, it should use that version. But it seems your environment is tricky for rubygems/bundler.

**Comment by sffc** [comment]
I encountered this error, too.  I replied to the StackOverflow question:

https://stackoverflow.com/a/50695884/1407170

In my case, the issue was that the binaries in image_optim_pack required a newer version of glibc than was available on my system.  My workaround (short up upgrading my Linux distribution, which I plan to do soon) was to install the latest version of image_optim_pack that worked, which appeared to be version 0.2.3.

**Comment by toy** [comment]
@sffc Last compatible version should be [0.3.1.20170318](https://github.com/toy/image_optim_pack/tree/v0.3.1.20170318) as I've switched from centos to ubuntu in be36a52adf8ad7b1066076c176f518eb66dd0345 and rebuilt all binaries in 0cdae11c2bdf061b839d40a8169635abe64acdfd before releasing [0.4.0](https://github.com/toy/image_optim_pack/tree/v0.4.0). I'll check what can be done about required glibc version.

**Comment by toy** [comment]
@sffc Please try version on branch [centos-6.9](https://github.com/toy/image_optim_pack/tree/centos-6.9).

**Comment by sffc** [comment]
I put this in my Gemfile and then ran bundle install:

```ruby
gem "image_optim_pack", :git => "git://github.com/toy/image_optim_pack.git", :branch => "centos-6.9"
```

Output of `bundle exec image_optim --info`:

```
image_optim v0.26.1
config:
  verbose: true
nice: 10
threads: 4
pack: true
skip_missing_workers: true
allow_lossy: false
cache_dir:
cache_worker_digests: false
image_optim_pack: all bins from /path/to/.rvm/gems/ruby-2.4.0/bundler/gems/image_optim_pack-f868817156b1/vendor/linux-x86_64 worked
Resolved pngcrush 1.8.13 at /path/to/.rvm/gems/ruby-2.4.0/bundler/gems/image_optim_pack-f868817156b1/vendor/linux-x86_64/pngcrush
Resolved advpng 2.1 at /path/to/.rvm/gems/ruby-2.4.0/bundler/gems/image_optim_pack-f868817156b1/vendor/linux-x86_64/advpng
Resolved optipng 0.7.7 at /path/to/.rvm/gems/ruby-2.4.0/bundler/gems/image_optim_pack-f868817156b1/vendor/linux-x86_64/optipng
Resolved pngquant 2.12.0 at /path/to/.rvm/gems/ruby-2.4.0/bundler/gems/image_optim_pack-f868817156b1/vendor/linux-x86_64/pngquant
Resolved jhead 3.00 at /path/to/.rvm/gems/ruby-2.4.0/bundler/gems/image_optim_pack-f868817156b1/vendor/linux-x86_64/jhead
Resolved jpegtran 9c at /path/to/.rvm/gems/ruby-2.4.0/bundler/gems/image_optim_pack-f868817156b1/vendor/linux-x86_64/jpegtran
Resolved jpegoptim 1.4.6 at /path/to/.rvm/gems/ruby-2.4.0/bundler/gems/image_optim_pack-f868817156b1/vendor/linux-x86_64/jpegoptim
Resolved jpegrescan 1a762f62 at /path/to/.rvm/gems/ruby-2.4.0/gems/image_optim-0.26.1/vendor/jpegrescan
Resolved gifsicle 1.91 at /path/to/.rvm/gems/ruby-2.4.0/bundler/gems/image_optim_pack-f868817156b1/vendor/linux-x86_64/gifsicle
pngout worker: `pngout` not found; please provide proper binary or disable this worker (--no-pngout argument or `:pngout => false` through options)
svgo worker: `svgo` not found; please provide proper binary or disable this worker (--no-svgo argument or `:svgo => false` through options)
Workers by format:
( ... )
```

So it looks like the branch works. 😃

**Comment by toy** [comment]
@sffc v0.5.1 is out
